### ISSUE TITLE
Vagrant claims script finishes with non-zero exit status when it doesn't

### ISSUE THREAD

**Issue by iwishiwasaneagle** [hashicorp/vagrant#11715]
### Vagrant version
`Vagrant 2.2.9`

### Host operating system
`$ uname -a`
`5.6.16-1-MANJARO` 

### Guest operating system
ubuntu/bionic64`

### Vagrantfile

```ruby
# -*- mode: ruby -*-
## vi: set ft=ruby :

VAGRANTFILE_API_VERSION = "2"

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.ssh.forward_x11 = true

  config.vm.provider "virtualbox" do |vb|
      # Don't boot with headless mode
      #   vb.gui = true
      #
      #   # Use VBoxManage to customize the VM. For example to change memory:
      vb.customize ["modifyvm", :id, "--memory", "8192"] # Out of 16GB
      vb.customize ["modifyvm", :id, "--ioapic", "on"]
      vb.customize ["modifyvm", :id, "--cpus", "10"] # Out of 16
      vb.customize ["modifyvm", :id, "--cpuexecutioncap", "70"] # Out of 100%
      # Make some effort to avoid clock skew
      vb.customize ["guestproperty", "set", :id, "/VirtualBox/GuestAdd/VBoxService/--timesync-set-threshold", "5000"]
      vb.customize ["guestproperty", "set", :id, "/VirtualBox/GuestAdd/VBoxService/--timesync-set-start"]
      vb.customize ["guestproperty", "set", :id, "/VirtualBox/GuestAdd/VBoxService/--timesync-set-on-restore", "1"]
  end

  # 18.04 LTS
  config.vm.define "bionic64", primary: true do |bionic64|
    bionic64.vm.box = "ubuntu/bionic64"

    # Setting up dev environment
    bionic64.vm.provision :shell, path: "scripts/setup_env.sh", privileged: false

    # Installing ROS, MAVROS, PX4, Gazebo 9, jMAVSim
    bionic64.vm.provision :shell, path: "scripts/init_ws.sh", privileged: false, keep_color:true, name: "Init workspace"

    bionic64.vm.provider "virtualbox" do |vb|
      vb.name = "PX4 (bionic64)"
    end
  end
end

```

### Debug output

`vagrant up --debug 2>&1 | tee debug.log`
https://gist.github.com/iwishiwasaneagle/635b872d06eb3d26fb7a230a993fe889

### Expected behavior

Clean exit of `vagrant up`

### Actual behavior

vagrant reports that ssh finished with a non-zero exit status. I know it finishes with a 0 exit code, as I force it within the script after seeing this issue.

### Steps to reproduce

1. run the `Vagrantfile` for the first time (or after `vagrant destroy`)

**Comment by jbonhag** [comment]
Hi there,

Thanks for opening an issue with Vagrant. Would you mind running the script manually inside the guest and checking the exit code?

For example:

```shell
$ vagrant up --no-provision
$ vagrant ssh
vagrant@ubuntu-bionic:~$ bash /vagrant/scripts/setup_env.sh
vagrant@ubuntu-bionic:~$ echo $?
```

I also noticed an error related to building '/home/vagrant/Fast-RTPS-Gen/build.gradle' in the debug logs. Could that cause a non-zero exit?

Cheers!

**Comment by jbonhag** [comment]
Sorry, please run `/vagrant/scripts/init_ws.sh` since that's the one we're debugging. Thanks!

**Comment by iwishiwasaneagle** [comment]
Hi @jbonhag , thanks for taking the time to help me out.

```bash
vagrant@ubuntu-bionic:~$ echo $?
0
```

Please note that I have the `set -e` flag yet the script runs all the way to my `echo "Success"` debug statement.

**Comment by chrisroberts** [comment]
@iwishiwasaneagle Would you run the `vagrant up` again but use the `--debug-timestamp` flag instead and provide a gist to that. Thanks!

**Comment by iwishiwasaneagle** [comment]
@chrisroberts 
`$ vagrant up --debug-timestamp 2>&1 | tee /tmp/vagrant.log`
[vagrant.log](https://github.com/hashicorp/vagrant/files/4826551/vagrant.log)

**Comment by chrisroberts** [comment]
@iwishiwasaneagle Thanks. This isn't quite what I need but my apologies, I wasn't very clear in what I was asking for. Can you first destroy the guest, and then run `vagrant up --debug-timestamp`? I'd like to see the full log from creation, through provision, and the resulting error from the successful provision script. I realize this is what was provided originally, but the additional timing information will hopefully help to narrow down where exactly the error is originating from.

Thanks again!

**Comment by iwishiwasaneagle** [comment]
Hi @chrisroberts, apologies. Please find attached the log output from `vagrant destroy && vagrant -up --debug-timestamp 2>1 | tee /tmp/vagrant.log`
[vagrant.log](https://github.com/hashicorp/vagrant/files/4829981/vagrant.log)

**Comment by chrisroberts** [comment]
Fantastic, thanks!

**Comment by pasankg** [comment]
Hi,
I'm getting the same error as well. I've tried https://github.com/hashicorp/vagrant/issues/1659#issuecomment-17046154 but no luck. 

Vagrant 2.2.9
Virtualbox: 6.1.10 r138449 (Qt5.6.3)
Host: MacOs 10.15.5

Vagrant log: [https://gist.github.com/pasankg/c7fea5df203f1149bd6ce8e4da1ebc77](https://gist.github.com/pasankg/c7fea5df203f1149bd6ce8e4da1ebc77)
VagrantFile: [https://gist.github.com/pasankg/59a337e87cbc1f417cfed48cdbac1ade](https://gist.github.com/pasankg/59a337e87cbc1f417cfed48cdbac1ade)

Would really appreciate if someone could help.
Thank you

**Comment by jbonhag** [comment]
Hey @pasankg,

It looks like your error is caused by a problem mounting an NFS share:

```
DEBUG ssh: stderr: mount.nfs: Connection timed out

DEBUG ssh: Exit status: 32
ERROR warden: Error occurred: The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o vers=3 192.168.200.1:/System/Volumes/Data/Users/pasgamag/www/estate /var/www/estate
```

This is a slightly different problem than the original issue, where the SSH command is not receiving an exit status at all.

If you're still experiencing this issue, would you be willing to open a new issue with a minimal Vagrantfile that demonstrates the issue? Thanks!

**Comment by jbonhag** [comment]
HI @iwishiwasaneagle,

Thanks so much for the additional logs. We were able to determine that the SSH command never returns an exit status, and Vagrant interprets the nil exit status as a non-zero/failure. It may be that the guest operating system is terminating the provisioning script before it completes -- do you see any out-of-memory messages in the guest's system logs?

**Comment by iwishiwasaneagle** [comment]
> Hi @iwishiwasaneagle,
> 
> Thanks so much for the additional logs. We were able to determine that the SSH command never returns an exit status, and Vagrant interprets the nil exit status as a non-zero/failure. It may be that the guest operating system is terminating the provisioning script before it completes -- do you see any out-of-memory messages in the guest's system logs?

Hi @jbonhag, 

No I didn't see anything of the sort. However, I modified the init_ws.sh script and it resolved the issue. It seems the installation of ros melodic along with mavros was causing issues. This has been a problem for me in the past yet I don't see anything relevant in the logs. 

Here is the log for the scripts/init_ws.sh file. The issue was resolved after commit `251d0327f55ae20c0f332e93f99d7092945aaf46`.

```
commit c30996c95929c1a05175598fe41fd276b9d86539
Author: iwishiwasaneagle <jh.ewers@gmail.com>
Date:   Mon Jun 29 18:10:14 2020 +0100

    Github & some gstreamer dependencies

diff --git a/scripts/init_ws.sh b/scripts/init_ws.sh
index f5e5d69..294df65 100755
--- a/scripts/init_ws.sh
+++ b/scripts/init_ws.sh
@@ -66,8 +66,14 @@ sudo apt-get install -y libprotobuf-dev libprotoc-dev protobuf-compiler libeigen
 sudo apt-get install -y libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-bad gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-ugly 
 
 if [ ! -d ~/Firmware ]; then
+	cat /vagrant/keys/id_rsa_github > ~/.ssh/id_rsa
+	cat /vagrant/keys/id_rsa_pub_github > ~/.ssh/id_rsa.pub
+	chmod 0600 ~/.ssh/id_rsa
+	chmod 0644 ~/.ssh/id_rsa.pub
+	ssh-add ~/.ssh/id_rsa
 	cd ~/
-	git clone https://github.com/PX4/Firmware.git --recursive
+	git clone [HIDDEN] --recursive
+	cd Firmware
 fi
 
 # MAVSDK-Python

commit 251d0327f55ae20c0f332e93f99d7092945aaf46
Author: iwishiwasaneagle <jh.ewers@gmail.com>
Date:   Mon Jun 29 15:34:39 2020 +0000

    Removed ROS, add MAVSDK

diff --git a/scripts/init_ws.sh b/scripts/init_ws.sh
index ba3c22d..f5e5d69 100755
--- a/scripts/init_ws.sh
+++ b/scripts/init_ws.sh
@@ -1,8 +1,8 @@
 #!/bin/bash
 
 # Show every step, then execute. Break on error
-#set -o xtrace
-#set -e
+set -o xtrace
+set -e
 
 ## Bash script for setting up ROS Melodic (with Gazebo 9) development environment for PX4 on Ubuntu LTS (18.04).
 ## It installs the common dependencies for all targets (including Qt Creator)
@@ -14,21 +14,28 @@
 
 USER=vagrant
 
-if [[ $(lsb_release -sc) == *"xenial"* ]]; then
-  echo "OS version detected as $(lsb_release -sc) (16.04)."
-  echo "ROS Melodic requires at least Ubuntu 18.04."
-  echo "Exiting ...."
-  return 1;
-fi
-
-echo "Downloading dependent script 'ubuntu_sim_common_deps.sh'"
-# Source the ubuntu_sim_common_deps.sh script directly from github
-common_deps=$(wget https://raw.githubusercontent.com/PX4/Devguide/master/build_scripts/ubuntu_sim_common_deps.sh -O -)
-wget_return_code=$?
-# If there was an error downloading the dependent script, we must warn the user and exit at this point.
-if [[ $wget_return_code -ne 0 ]]; then echo "Error downloading 'ubuntu_sim_common_deps.sh'. Sorry but I cannot proceed further :("; exit 1; fi
-# Otherwise source the downloaded script.
-. <(echo "${common_deps}")
+# Ubuntu Config for PX4
+echo "Remove modemmanager"
+sudo apt-get remove modemmanager -y
+echo "Add user to dialout group for serial port access (reboot required)"
+sudo usermod -a -G dialout $USER
+
+# Common dependencies for PX4
+echo "Installing common dependencies"
+sudo apt-get update -y
+sudo apt-get install git zip cmake build-essential genromfs ninja-build exiftool astyle libopencv-dev -y
+# make sure xxd is installed, dedicated xxd package since Ubuntu 18.04 but was squashed into vim-common before
+which xxd || sudo apt install xxd -y || sudo apt-get install vim-common --no-install-recommends -y
+# Required python packages
+sudo apt-get install python3-dev python3-pip -y
+sudo -H pip3 install --upgrade pip
+sudo -H pip3 install pandas jinja2 pyserial pyyaml empy toml numpy pyros-genmsg packaging
+# optional python tools
+sudo -H pip3 install pyulog
+
+# jMAVSim simulator dependencies
+echo "Installing jMAVSim simulator dependencies"
+sudo apt-get install ant openjdk-8-jdk openjdk-8-jre -y
 
 # FastRTPS
 fastrtps_dir=~/FastRTPS-1.8.2
@@ -51,86 +58,10 @@ else
     sudo ln -sf /home/$USER/Fast-RTPS-Gen/scripts/fastrtpsgen /usr/local/bin/
 fi
 
-# ROS Melodic
-## Gazebo simulator dependencies
-sudo apt-get install protobuf-compiler libeigen3-dev libopencv-dev -y
-
-## ROS Gazebo: http://wiki.ros.org/melodic/Installation/Ubuntu
-## Setup keys
-sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
-sudo apt-key adv --keyserver 'hkp://keyserver.ubuntu.com:80' --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
-## For keyserver connection problems substitute hkp://pgp.mit.edu:80 or hkp://keyserver.ubuntu.com:80 above.
-sudo apt-get update
-## Get ROS/Gazebo
-
-# Dependencies
-sudo apt install ros-melodic-desktop-full ros-melodic-message-generation ros-melodic-message-runtime libgeographic-dev ros-melodic-geographic-msgs \
-		 python-wstool build-essential \
-		 python-catkin-tools python-future python-lxml -y
-
-sudo apt install -y python-pip
-sudo pip install -U rosdep rosinstall_generator vcstool rosinstall
-## Initialize rosdep
-sudo rosdep init
-rosdep update
-## Setup environment variables
-rossource="source /opt/ros/melodic/setup"
-if grep -Fxq "$rossource.zsh" ~/.zshrc; then echo ROS setup.zsh already in .zshrc;
-else echo "$rossource.zsh" >> ~/.zshrc; fi
-eval "$rossource.bash"
-
-# MAVROS: https://dev.px4.io/en/ros/mavros_installation.html
-## Create catkin workspace
-mkdir -p ~/catkin_ws/src
-cd ~/catkin_ws
-catkin init
-wstool init src
-
-
-## Install MAVLink
-###we use the Kinetic reference for all ROS distros as it's not distro-specific and up to date
-rosinstall_generator --rosdistro melodic mavlink | tee /tmp/mavros.rosinstall
-
-## Build MAVROS
-### Get source (upstream - released)
-rosinstall_generator --upstream mavros | tee -a /tmp/mavros.rosinstall
-
-### Setup workspace & install deps
-wstool merge -t src /tmp/mavros.rosinstall
-wstool update -t src
-if ! rosdep install --from-paths src --ignore-src -y; then
-    # (Use echo to trim leading/trailing whitespaces from the unsupported OS name
-    unsupported_os=$(echo $(rosdep db 2>&1| grep Unsupported | awk -F: '{print $2}'))
-    rosdep install --from-paths src --ignore-src --rosdistro melodic -y --os ubuntu:bionic
-fi
-
-if [[ ! -z $unsupported_os ]]; then
-    >&2 echo -e "\033[31mYour OS ($unsupported_os) is unsupported. Assumed an Ubuntu 18.04 installation,"
-    >&2 echo -e "and continued with the installation, but if things are not working as"
-    >&2 echo -e "expected you have been warned."
-fi
-
-#Install geographiclib
-sudo apt install geographiclib-tools -y
-echo "Downloading dependent script 'install_geographiclib_datasets.sh'"
-# Source the install_geographiclib_datasets.sh script directly from github
-install_geo=$(wget https://raw.githubusercontent.com/mavlink/mavros/master/mavros/scripts/install_geographiclib_datasets.sh -O -)
-wget_return_code=$?
-# If there was an error downloading the dependent script, we must warn the user and exit at this point.
-if [[ $wget_return_code -ne 0 ]]; then echo "Error downloading 'install_geographiclib_datasets.sh'. Sorry but I cannot proceed further :("; exit 1; fi
-# Otherwise source the downloaded script.
-sudo bash -c "$install_geo"
+# Gazebo 11.0
+curl -sSL http://get.gazebosim.org | sh
 
-## Build!
-catkin build
-## Re-source environment to reflect new packages/build environment
-catkin_ws_source="source ~/catkin_ws/devel/setup"
-if grep -Fxq "$catkin_ws_source.zsh" ~/.zshrc; then echo ROS catkin_ws setup.zsh already in .zshrc;
-else echo "$catkin_ws_source.zsh" >> ~/.zshrc; fi
-eval "$catkin_ws_source.bash"
-
-# PX4 SITL
-sudo apt-get install -y python3-empy python3-toml python3-packaging python3-numpy
+# PX4 SITL Dependencies
 sudo apt-get install -y libprotobuf-dev libprotoc-dev protobuf-compiler libeigen3-dev libxml2-utils python-rospkg python-jinja2
 sudo apt-get install -y libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-bad gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-ugly 
 
@@ -138,3 +69,6 @@ if [ ! -d ~/Firmware ]; then
 	cd ~/
 	git clone https://github.com/PX4/Firmware.git --recursive
 fi
+
+# MAVSDK-Python
+sudo -H pip3 install mavsdk

commit fb11e524f8ea31ee7e3fc546691bcaa058887297
Author: iwishiwasaneagle <jh.ewers@gmail.com>
Date:   Thu Jun 25 12:34:58 2020 +0100

    Vagrant

diff --git a/scripts/init_ws.sh b/scripts/init_ws.sh
new file mode 100755
index 0000000..ba3c22d
--- /dev/null
+++ b/scripts/init_ws.sh
@@ -0,0 +1,140 @@
+#!/bin/bash
+
+# Show every step, then execute. Break on error
+#set -o xtrace
+#set -e
+
+## Bash script for setting up ROS Melodic (with Gazebo 9) development environment for PX4 on Ubuntu LTS (18.04).
+## It installs the common dependencies for all targets (including Qt Creator)
+##
+## Installs:
+## - Common dependencies libraries and tools as defined in `ubuntu_sim_common_deps.sh`
+## - ROS Melodic (including Gazebo9)
+## - MAVROS
+
+USER=vagrant
+
+if [[ $(lsb_release -sc) == *"xenial"* ]]; then
+  echo "OS version detected as $(lsb_release -sc) (16.04)."
+  echo "ROS Melodic requires at least Ubuntu 18.04."
+  echo "Exiting ...."
+  return 1;
+fi
+
+echo "Downloading dependent script 'ubuntu_sim_common_deps.sh'"
+# Source the ubuntu_sim_common_deps.sh script directly from github
+common_deps=$(wget https://raw.githubusercontent.com/PX4/Devguide/master/build_scripts/ubuntu_sim_common_deps.sh -O -)
+wget_return_code=$?
+# If there was an error downloading the dependent script, we must warn the user and exit at this point.
+if [[ $wget_return_code -ne 0 ]]; then echo "Error downloading 'ubuntu_sim_common_deps.sh'. Sorry but I cannot proceed further :("; exit 1; fi
+# Otherwise source the downloaded script.
+. <(echo "${common_deps}")
+
+# FastRTPS
+fastrtps_dir=~/FastRTPS-1.8.2
+echo "Installing FastRTPS to: $fastrtps_dir"
+if [ -d "$fastrtps_dir" ]
+then
+    echo " FastRTPS already installed."
+else
+    sudo apt-get install -y gradle libssl-dev
+    git clone --recursive https://github.com/eProsima/Fast-RTPS.git -b 1.8.x $fastrtps_dir
+    cd ~/FastRTPS-1.8.2
+    mkdir build && cd build
+    cmake -DTHIRDPARTY=ON -DSECURITY=ON ..
+    make
+    sudo make install
+    git clone --recursive https://github.com/eProsima/Fast-RTPS-Gen.git -b v1.0.4 ~/Fast-RTPS-Gen \
+        && cd ~/Fast-RTPS-Gen \
+        && gradle assemble \
+        && gradle install
+    sudo ln -sf /home/$USER/Fast-RTPS-Gen/scripts/fastrtpsgen /usr/local/bin/
+fi
+
+# ROS Melodic
+## Gazebo simulator dependencies
+sudo apt-get install protobuf-compiler libeigen3-dev libopencv-dev -y
+
+## ROS Gazebo: http://wiki.ros.org/melodic/Installation/Ubuntu
+## Setup keys
+sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
+sudo apt-key adv --keyserver 'hkp://keyserver.ubuntu.com:80' --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
+## For keyserver connection problems substitute hkp://pgp.mit.edu:80 or hkp://keyserver.ubuntu.com:80 above.
+sudo apt-get update
+## Get ROS/Gazebo
+
+# Dependencies
+sudo apt install ros-melodic-desktop-full ros-melodic-message-generation ros-melodic-message-runtime libgeographic-dev ros-melodic-geographic-msgs \
+		 python-wstool build-essential \
+		 python-catkin-tools python-future python-lxml -y
+
+sudo apt install -y python-pip
+sudo pip install -U rosdep rosinstall_generator vcstool rosinstall
+## Initialize rosdep
+sudo rosdep init
+rosdep update
+## Setup environment variables
+rossource="source /opt/ros/melodic/setup"
+if grep -Fxq "$rossource.zsh" ~/.zshrc; then echo ROS setup.zsh already in .zshrc;
+else echo "$rossource.zsh" >> ~/.zshrc; fi
+eval "$rossource.bash"
+
+# MAVROS: https://dev.px4.io/en/ros/mavros_installation.html
+## Create catkin workspace
+mkdir -p ~/catkin_ws/src
+cd ~/catkin_ws
+catkin init
+wstool init src
+
+
+## Install MAVLink
+###we use the Kinetic reference for all ROS distros as it's not distro-specific and up to date
+rosinstall_generator --rosdistro melodic mavlink | tee /tmp/mavros.rosinstall
+
+## Build MAVROS
+### Get source (upstream - released)
+rosinstall_generator --upstream mavros | tee -a /tmp/mavros.rosinstall
+
+### Setup workspace & install deps
+wstool merge -t src /tmp/mavros.rosinstall
+wstool update -t src
+if ! rosdep install --from-paths src --ignore-src -y; then
+    # (Use echo to trim leading/trailing whitespaces from the unsupported OS name
+    unsupported_os=$(echo $(rosdep db 2>&1| grep Unsupported | awk -F: '{print $2}'))
+    rosdep install --from-paths src --ignore-src --rosdistro melodic -y --os ubuntu:bionic
+fi
+
+if [[ ! -z $unsupported_os ]]; then
+    >&2 echo -e "\033[31mYour OS ($unsupported_os) is unsupported. Assumed an Ubuntu 18.04 installation,"
+    >&2 echo -e "and continued with the installation, but if things are not working as"
+    >&2 echo -e "expected you have been warned."
+fi
+
+#Install geographiclib
+sudo apt install geographiclib-tools -y
+echo "Downloading dependent script 'install_geographiclib_datasets.sh'"
+# Source the install_geographiclib_datasets.sh script directly from github
+install_geo=$(wget https://raw.githubusercontent.com/mavlink/mavros/master/mavros/scripts/install_geographiclib_datasets.sh -O -)
+wget_return_code=$?
+# If there was an error downloading the dependent script, we must warn the user and exit at this point.
+if [[ $wget_return_code -ne 0 ]]; then echo "Error downloading 'install_geographiclib_datasets.sh'. Sorry but I cannot proceed further :("; exit 1; fi
+# Otherwise source the downloaded script.
+sudo bash -c "$install_geo"
+
+## Build!
+catkin build
+## Re-source environment to reflect new packages/build environment
+catkin_ws_source="source ~/catkin_ws/devel/setup"
+if grep -Fxq "$catkin_ws_source.zsh" ~/.zshrc; then echo ROS catkin_ws setup.zsh already in .zshrc;
+else echo "$catkin_ws_source.zsh" >> ~/.zshrc; fi
+eval "$catkin_ws_source.bash"
+
+# PX4 SITL
+sudo apt-get install -y python3-empy python3-toml python3-packaging python3-numpy
+sudo apt-get install -y libprotobuf-dev libprotoc-dev protobuf-compiler libeigen3-dev libxml2-utils python-rospkg python-jinja2
+sudo apt-get install -y libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-bad gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-ugly 
+
+if [ ! -d ~/Firmware ]; then
+	cd ~/
+	git clone https://github.com/PX4/Firmware.git --recursive
+fi

```
